pkg/auth/oauth/external/keycloak/keycloak.go

```diff
@@ -0,0 +1,132 @@
+package keycloak
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/RangelReale/osincli"
+	"github.com/golang/glog"
+
+	authapi "github.com/openshift/origin/pkg/auth/api"
+	"github.com/openshift/origin/pkg/auth/oauth/external"
+)
+
+type provider struct {
+	Realm          string `json:"realm"`
+	RealmPublicKey string `json:"realm-public-key"`
+	ClientID       string `json:"resource"`
+	Credentials    struct {
+		Secret string `json:"secret"`
+	} `json:"credentials"`
+	AuthServerURL string `json:"auth-server-url"`
+}
+
+type keycloakUser struct {
+	id         string
+	username   string
+	email      string
+	firstName  string
+	lastName   string
+	realmRoles string
+}
+
+func NewProvider(keycloakClientConfigFile string) external.Provider {
+	keycloakClientConfigBytes, err := ioutil.ReadFile(keycloakClientConfigFile)
+	if err != nil {
+		glog.Errorf("Error loading Keycloak config: %s", err)
+		return nil
+	}
+	p := provider{}
+	err = json.Unmarshal(keycloakClientConfigBytes, &p)
+	if err != nil {
+		glog.Errorf("Error parsing Keycloak config: %s", err)
+		return nil
+	}
+	return p
+}
+
+// NewConfig implements external/interfaces/Provider.NewConfig
+func (p provider) NewConfig() (*osincli.ClientConfig, error) {
+	glog.V(4).Infof("Keycloak provider=%v", p)
+	config := &osincli.ClientConfig{
+		ClientId:                 p.ClientID,
+		ClientSecret:             p.Credentials.Secret,
+		ErrorsInStatusCode:       true,
+		SendClientSecretInParams: true,
+		AuthorizeUrl:             p.AuthServerURL + "/realms/" + p.Realm + "/tokens/login",
+		TokenUrl:                 p.AuthServerURL + "/realms/" + p.Realm + "/tokens/access/codes",
+	}
+	return config, nil
+}
+
+// AddCustomParameters implements external/interfaces/Provider.AddCustomParameters
+func (p provider) AddCustomParameters(req *osincli.AuthorizeRequest) {
+}
+
+// GetUserIdentity implements external/interfaces/Provider.GetUserIdentity
+func (p provider) GetUserIdentity(data *osincli.AccessData) (authapi.UserIdentityInfo, bool, error) {
+	idToken, ok := data.ResponseData["id_token"].(string)
+	if !ok {
+		return nil, false, fmt.Errorf("No id_token returned in %v", data.ResponseData)
+	}
+
+	userdata, err := decodeJWT(idToken)
+	glog.V(4).Infof("userdata=%v", userdata)
+	if err != nil {
+		return nil, false, err
+	}
+
+	username, _ := userdata["preferred_username"].(string)
```

Comment:
is this mutable? we want to make sure the value returned from the identity server is one we can count on not changing